posts/2018-11-28-http4s-error-handling-mtl-2.md

```diff
@@ -0,0 +1,459 @@
+
+#### Data types with error handling capabilities
+
+As of now I still prefer the latter for a number of reasons. Here's a comparison table of all the different approaches I can think of:
+
+|            | EitherT[IO, E, ?] | IO[E, A] | IO[A] | IO[A] + Classy Optics |
+|------------|-------------------|----------|-------|-----------------------|
+| Concise code style | ✘ | ✔ | ✔ | ✔ |
+| Polymorphic code | ✔ | ✘ | ✔ | ✔ |
+| Performance overhead | **2x** | ✔ | ✔ | ✔ |
+| Double error channel | ✔ | ✔ | ✘ | ✘ |
+
+With "Double error channel" I mean that we can signal failure in two different ways:
+
+- `EitherT[IO, E, ?]` can either fail with `Left` or with `IO.raiseError` (that can also be a caught exception).
+- `IO[E, A]` can either fail with an `E` or by throwing an exception when `A` was expected. This is called "unrecoverable" errors.
+
+And this is not a bad thing at all! It is like this by design and we can have the same property when writing polymorphic code using `cats-effect` while still keeping it simple. Here's one way:
+
+### Error Channel
+
+In the previous blog post we defined the algebras as a single trait. In this case we are going to try a different encoding but first we need to introduce an `ErrorChannel[F, E]` typeclass where the error type is a subtype of `Throwable` to be compatible with the error type of the `cats-effect` typeclasses:
+
+```tut:book:silent
+trait ErrorChannel[F[_], E <: Throwable] {
+  def raise[A](e: E): F[A]
+}
+```
+
+An instance can be derived for any `ApplicativeError[F, Throwable]` so we don't need to write it manually for every error type.
+
+```tut:book:silent
+import cats.ApplicativeError
+
+object ErrorChannel {
+  def apply[F[_], E <: Throwable](implicit ev: ErrorChannel[F, E]) = ev
+
+  implicit def instance[F[_], E <: Throwable](implicit F: ApplicativeError[F, Throwable]): ErrorChannel[F, E] =
+    new ErrorChannel[F, E] {
+      override def raise[A](e: E) = F.raiseError(e)
+    }
+
+  object syntax {
+    implicit class ErrorChannelOps[F[_]: ErrorChannel[?[_], E], E <: Throwable](e: E) {
+      def raise[A]: F[A] = ErrorChannel[F, E].raise[A](e)
+    }
+  }
+}
+```
+
+### User Algebra
+
+Our `UserAlg` will now be defined as an `abstract class` instead in order to be able to add typeclass constraint.
+
+```tut:book:silent
+case class User(username: String, age: Int)
+case class UserUpdateAge(age: Int)
+
+abstract class UserAlg[F[_]: ErrorChannel[?[_], E], E <: Throwable] {
+  def find(username: String): F[Option[User]]
+  def save(user: User): F[Unit]
+  def updateAge(username: String, age: Int): F[Unit]
+}
+```
+
+And here's the ADT of the possible errors that may arise (notice the `extends Exception` part):
```

Comment:
We're assuming the entire audience knows that an `Exception` is a `Throwable`?

Reply:
Added a comment below clarifying this.